proposed/0006-Launcher.md

```diff
@@ -0,0 +1,192 @@
+
+**step 3:** Launcher checks the App-ID, reads the path from the `<LOCAL-CONFIG-FILE>` that it made and starts the app as an independent process. The Launcher supplies a random port on which it will listen to this app via command line options.
+
+**./path/to/XYZ --launcher “port:33000;protocol:udp;nonce:random-u32-number”**
+
+**step 4:** Launcher will wait for a predefined time of **15** seconds for data reception on that port. If it times out it will close the socket (release its binding to it)
+
+**step 5:** App responds on the socket asking for Launcher to give keys and its root directory, which Launcher had reserved as `XYZ-Root-Dir`
+- The payload format for this request shall be CBOR encoded sturcture of the following:
+```
+struct Request {
+    prefix: u8,  // This must be ASCII of '?' (a question mark)
+    nonce : u32, // This must be the nonce supplied by Launcher
+}
+```
+
+**step 6:** Launcher reads the `<LAUNCHER-CONFIG-FILE>` and creates a mapping in the MaidManagers for associating user’s `MAID Keys <-> App specific Keys` to allow App to PUT/POST on behalf of the user in `<APP-ROOT-DIR>`.
+- The mapping is done only for the ownership key/s (not encryption keys).
+- The request for mapping/un-mapping shall be a command to the `MaidManagers`. For commands to `MaidManagers`, `StructuredData` with special Type-Tag will be reserved. On reception of this `StructuredData` the vaults will check the payload and act upon the request instead of executing a normal reaction to the usual `PUT/DELETES`.
+- `StructuredData` with Tag-Type **9** will be reserved for `Client <-> Vault` messages.
+- The messages will be defined in [safe_vault](https://github.com/maidsafe/safe_vault).
+- This shall be the format of `Client <-> Vault` messages:
+```
+pub enum ClientVaultMessage {
+    MapKeys {
+        ownership_key_to_map   : sodiumoxide::crypto::sign::PublicKey, // Launcher-Assigned-Public-Key-For-App
+        ownership_key_to_map_to: sodiumoxide::crypto::sign::PublicKey, // User-PublicMaid-sign-key
+    },
+    UnMapKeys {
+        ownership_key_to_unmap     : sodiumoxide::crypto::sign::PublicKey, // Launcher-Assigned-Public-Key-For-App
+        ownership_key_to_unmap_from: sodiumoxide::crypto::sign::PublicKey, // User-PublicMaid-sign-key
+    },
+}
+```
+
+**step 7:** Launcher gives the **App Keys** and **App-Dir** to the App and at this point it closes its socket and is no longer associated with the App in anyway.
+- The payload format for this response shall be CBOR encoded sturcture of the following, followed by hybrid_encrypt of the stream using App-specific-crypto-keys:
+```
+struct Response {
+    root_directory_id    : NameType,
+    root_directory_tag   : u64
+    public_signing_key   : sodiumoxide::crypto::sign::PublicKey,
+    private_signing_key  : sodiumoxide::crypto::sign::SecretKey,
+    public_encrytion_key : sodiumoxide::crypto::box_::PublicKey,
+    private_encrytion_key: sodiumoxide::crypto::box_::SecretKey,
+}
+```
+
```

Comment:
We had planned this before to have the app generate its own keys and pass them to the Launcher.  This differed in two major ways.
1. The Launcher never gets to see the app's private keys and they're never transmitted between these processes.  They're not stored in any config file anywhere - neither by the Launcher, nor the app.
2. The keys then are session keys.  They only need to be passed to the Vault network by the Launcher - not deleted.  The app itself can connect to the network at the same location as the Launcher, and if the Launcher has provided the app keys, the app can PUT/POST/DELETE.  Once the app disconnects from the network, the Vaults holding the app keys will know about this without needing to be told, and will remove them.

Having these as session keys is useful not only because it limits the time for which they are usable, but also it acts as a disincentive for the app to try and store them somewhere (e.g. unencrypted on the native filesystem).  It also dissociates the App and Launcher even further, since the Launcher doesn't have to be running in order for these app keys to be removed from the network.